### PR TITLE
[#11015] Break down question result fetch by section for scalability

### LINKED ISSUE #11015: Optimize frontend result page fetching
Part of #11011 

**Description of feature/enhancement**

- [ ] Utilize the updated backend endpoint to achieve loading response by section and question
- [ ] Replace spinner with progress bar

**Justification**

Loading by section and question mitigates the backend server load in large courses.
In the case of long response loading, we would need to show the user there has been progress made.

### PR DESCRIPTION
**Pull request by moziliar**
Fixes #11015 

**Outline of Solution**

Using the existing group by section toolkit, this PR breaks down per question result fetch into one or more result fetch by question and section.

By the nature of the section size limit, we can upper bound the sheer number of responses fetched per api call, therefore mitigating the OOM and GAE timeout issue caused by large courses where a question could consist of too many responses.

Pending stress test to confirm if this solution solves **any** reasonably sized question response set per section. The prime factor for such consideration is the size of the response (say, essay question which consists of 500 words minimally).

**Update**

Also initiates the splitting of backend response fetch by giver and receiver. A trichotomy enum `ResultFetchType` along the data path is used as a common object propagated along UI, Logic and Storage layers for a cleaner look.

The rationale behind the backend changes is to upper bound the number of responses fetched via this way, utilizing the section size limit of 100. By allowing fetching by section but for both giver and receiver, the size of the return could be higher considering a chained fetch.

### CODE REVIEW

**Review by daongochieu2810 — APPROVED** [review]
lgtm!

**Review by teikjun — COMMENTED** [review]
The code looks good to me! 👍 
Do you know what might be causing the e2e tests to fail? I've checked out the error logs but I'm still unsure about how we can fix it. 

Anyways, I'm currently adding the circular progress indicator (as planned in https://github.com/TEAMMATES/teammates/issues/11015) - will make a PR against this branch soon.

**Comment by moziliar** [comment]
Thanks @daongochieu2810 @teikjun !

I am still investigating the e2e failure too. It will take quite awhile unfortunately. Will update again if there's anything to change in order to pass the test.

**Comment by moziliar** [comment]
@teikjun i've identified the problem and it appears that the problem also involves the unfinished Question Stats type definition in the frontend. @wkurniawan07 I see that you left the TODO [there](https://github.com/TEAMMATES/teammates/blob/39dae8c5a76c76decf966a9aab063caf0122b1b6/src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts#L59), should we tackle that first before carrying on with this PR?

The issue with the current code is that it naively overwrites the current question's data with each section's data; it should be an `append` operation that incrementally adds in the data of one question from each sections; the blocker here appears to be statistics which is in the form of string json now.

**Comment by wkurniawan07** [comment]
I highly doubt it's the issue. The only time where statistics is obtained from back-end is when the question type is contribution. All other statistics are calculated on demand based on available data.

**Comment by moziliar** [comment]
@wkurniawan07 yup. Do you think I can just jsonify the string and perform concat before throwing it back as a string as a temporary measure?

File: src/main/java/teammates/common/datatransfer/ResultFetchType.java
Comment (hcwong) [inline src/main/java/teammates/common/datatransfer/ResultFetchType.java:4]: Nit but can I trouble you to add in "feedback" in front of giver and receiver here just for more clarity when viewing this file in isolation

**Review by hcwong — APPROVED** [review]
code lgtm :)

File: src/main/java/teammates/common/datatransfer/ResultFetchType.java
Comment (moziliar) [reply src/main/java/teammates/common/datatransfer/ResultFetchType.java:4]: Thanks for your insight! I am wondering if this enum can be used for general purpose and hence left out the feedback part here. In the meantime I will just add it in!

File: src/main/java/teammates/common/datatransfer/ResultFetchType.java
Comment (tiuweehan) [inline src/main/java/teammates/common/datatransfer/ResultFetchType.java:18]: Just a suggestion, perhaps you could add 2 boolean fields to this constructor, e.g. `fetchByGiver` and `fetchByReceiver`, and implement 2 methods `shouldFetchByGiver()` and `shouldFetchByReceiver()`. This is similar to `FeedbackQuestionType`.

This is so that later on, rather than implementing the same set of methods for both giver and receiver (e.g. `getFeedbackResponseCommentEntitiesForQuestionInReceiverSection`), you can just use a single method (e.g. `getFeedbackResponseCommentEntitiesForQuestion`) and modify its behaviour based on what is returned from `shouldFetchByGiver()` and `shouldFetchByReceiver()`.

**Review by tiuweehan — APPROVED** [review]
Overall lgtm! I left a suggestion in terms of how I feel the code could be structured, but it shouldn't affect the underlying logic of your changes.

File: src/main/java/teammates/common/datatransfer/ResultFetchType.java
Comment (moziliar) [reply src/main/java/teammates/common/datatransfer/ResultFetchType.java:18]: I see, this sounds quite extensible! I do have the same concern here but your way is much clearer :)
In some previous iterations of the review, @rrtheonlyone mentioned keeping things separated and hence there are only separate methods `getFeedbackResponseCommentEntitiesForQuestionInReceiverSection` and its giver counter part remaining; this was done intentionally after reverting a change that rope the two into one.

Comment (moziliar) [reply src/main/java/teammates/common/datatransfer/ResultFetchType.java:18]: Actually, after looking at some code segments, I think leaving the trichotomy as three different distinct unit methods (`InGiverSection`, `InReceiverSection`, `In(Both)Section` might be cleaner at the `Storage` layer, as then only `Logic` needs to figure out which one of the above methods to use to get the correct response needed.

Can clarify too if you meant to say `BOTH(Boolean fetchByGiver, Boolean fetchByReceiver)`?

**Comment by moziliar** [comment]
@rrtheonlyone the caveats in splitting fetch by giver section and receiver section is that the missing responses will overwrite the response at some point, whereas the missing responses are still needed for frontend display. The logic now separate them and aggregate both to form the final response display.

**Comment by moziliar** [comment]
This issue can likely solve #11070; to monitor the number of errors from instructor session result workflow to determine the efficacy of section limit on workload control, and apply the same to student record page workflow too.

**Review by teikjun — APPROVED** [review]
There's a merge conflict here, LGTM otherwise!

File: src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts
Comment (rrtheonlyone) [inline src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts:291]: We should not be fetching section names every-time the question tab is pressed. This should be a one-time operation.

File: src/web/app/components/question-types/question-statistics/question-statistics.ts
Comment (rrtheonlyone) [inline src/web/app/components/question-types/question-statistics/question-statistics.ts:46]: I am wondering how this still works... right now the stats calculation is done on the backend and it looks like the method relies on all data being there. Now that it is split by section/type, it feels like there is going to be some edge cases here

**Review by rrtheonlyone — CHANGES_REQUESTED** [review]
Almost there, solid work! 

Looked though the code - logic seems to be right and I think the overall implementation is clean 💯 Just some minor comments below

File: src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts
Comment (moziliar) [reply src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts:291]: So we should instead cache the section names and use it for the response fetching instead?

File: src/web/app/components/question-types/question-statistics/question-statistics.ts
Comment (moziliar) [reply src/web/app/components/question-types/question-statistics/question-statistics.ts:46]: At the moment, the backend will actually generate empty stats with invalid (e.g. -9999) data for the team members who have no response yet (including those whose entries are not yet populated). 

For example:
team 1 with Alice, Bob and Charlie; only Alice has responded:
Stats returned will be `{ Alice: <valid stats>, Bob: <invalid/placeholder stats>, Charlie: <invalid/placeholder stats>}`

This function is relying on that the backend continues to output in the same format. What do you think about this?

Comment (moziliar) [reply src/web/app/components/question-types/question-statistics/question-statistics.ts:46]: At the moment, the backend will actually generate empty stats with invalid (e.g. -9999) data for the team members who have no response yet (including those whose entries are not yet populated).

For example:
team 1 with Alice, Bob and Charlie; only Alice has responded:
Stats returned will be { Alice: <valid stats>, Bob: <invalid/placeholder stats>, Charlie: <invalid/placeholder stats>}

This function is relying on that the backend continues to output in the same format. What do you think about this?

File: src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts
Comment (rrtheonlyone) [reply src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts:291]: Yeah you have a point.. although it feels very unlikely that the section will change while this page is still open. 

It's the same logic as what if someone adds a new question while this page is open - the question will only be shown upon refresh. Plus remember, we only load the results once - you press the tab again and you will get the earlier results :)

The fix above works well for large courses but the tradeoff is the increased number of API calls. For small courses we are making an unusually high amount of API calls every time the button is pressed. I feel section names should be cached as it is a reasonable tradeoff.

Comment (moziliar) [reply src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts:291]: I see. So it shall be! I think the assumption that sections don't usually change is reasonably, given that it has no real life meaning to the instructor.

File: src/web/app/components/question-types/question-statistics/question-statistics.ts
Comment (moziliar) [reply src/web/app/components/question-types/question-statistics/question-statistics.ts:46]: Btw, the above observation was made after countless attempts trying to decipher the backend working of this part; the current logic is a bit arcane and front/backend data flow seems not very integral yet. I see TODO somewhere (about the stats struct which is currently still in a string form and requires jsonification), which might be the next goal to fix.

File: src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts
Comment (moziliar) [inline src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts:292]: @rrtheonlyone here the question loading will fail if the page fails to fetch sections on init; the workflow will prompt the user to refresh the page

Comment (rrtheonlyone) [reply src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts:292]: Yup that's how it works for any of the init fetches right?

**Comment by rrtheonlyone** [comment]
Also, I realised that splitting by giver&receiver section is going to be an unnecessary call for responses where the giver/recvr section is the same. We will be doing an unnecessary API call + db hit.

I feel in majority of the cases `giverSection` and `recipientSection` is going to be the same just by the distribution of questions in teammates (I would guess most feedback seems to be "Team" level). 

Can we clever about when we split based on question type?

File: src/web/app/components/question-types/question-statistics/question-statistics.ts
Comment (rrtheonlyone) [reply src/web/app/components/question-types/question-statistics/question-statistics.ts:46]: Yes it conveniently works out because this is a team level contribution. The data returned by backend will be correct for that section. We just need to keep track of the new data as you have done.

Also remove lines 48-50 as they do nothing.

**Comment by moziliar** [comment]
@rrtheonlyone in a sense yes; however, this is a correctness vs cost kind of balance. In terms of DB calls, you could see that the old way was also fetching both giver and receiver section's responses and de-duplicate (refer to some of the deleted lines in `src/main/java/teammates/storage/api/FeedbackResponsesDb.java`); this means that the number of DB calls remains the same before and after this change. As such, the x2 QPS problem seems to be the only compromise here.

**Comment by rrtheonlyone** [comment]
No it is fine then.. realised the db doesn't support OR queries. So, we are not incurring a significant cost by making 2 calls. No need for further investigation. An extra API call isn't as bad as an extra db read. 

Thanks for the quick response.

File: src/web/app/components/question-types/question-statistics/question-statistics.ts
Comment (moziliar) [reply src/web/app/components/question-types/question-statistics/question-statistics.ts:46]: Oh will fix that now...

**Comment by moziliar** [comment]
e2e test failure due to timezone sync in all 5 runs.

**Comment by moziliar** [comment]
@rrtheonlyone e2e passes after merging master. Ready for review

**Review by rrtheonlyone — CHANGES_REQUESTED** [review]
@moziliar Almost there, apart from just fixing the existing tests, we also need to add more tests to cover the new logic we have introduced. Some places that come to mind:
- `FeedbackSessionsLogicTest`
- `FeedbackResponsesLogicTest`
- `FeedbackResponseCommentsLogicTest`
- `FeedbackQuestionsLogicTest`
- `FeedbackResponsesDbTest`
- `FeedbackResponsesCommentsDbTest`

However, this might be a good first issue for newcomers. I will leave it to you @moziliar - you can create a new issue for the above / add the tests yourself. Once that is done, we are good to merge this one in!

**Comment by moziliar** [comment]
@rrtheonlyone thank you for your response!
I am wondering if they could become good second issue for the coming wave of contributors?

**Comment by rrtheonlyone** [comment]
@moziliar if you can settle now, please go ahead. It will be much easier for the person doing the change to write the tests!

There is work to be done and we can definitely create more issues for the next wave of contributors.

**Comment by wkurniawan07** [comment]
> we also need to add more tests to cover the new logic we have introduced

> this might be a good first issue for newcomers

Sorry, but this is not the right mindset. Change in code should be accompanied by change in test unless the fix is urgent, and this fix is not urgent.

**Comment by rrtheonlyone** [comment]
@moziliar could we get an ETA on this PR please? 

No particular rush but want to make sure that this is accounted for. It would be a shame if we had to drop it after a few months.

**Comment by halfwhole** [comment]
@moziliar kind reminder on this PR! how would you like to move forward with this?

**Comment by madanalogy** [comment]
@moziliar putting this PR on hold as there hasn't been any progress and it's blocking #11037. Once that PR is resolved you're welcome to continue working on this

**Comment by madanalogy** [comment]
Removing on hold status as #11037 closed due to inactivity

**Comment by daongochieu2810** [comment]
@moziliar kind reminder on this PR

**Comment by moziliar** [comment]
Will pick it up within the week. Thanks for the reminder.

**Comment by t-cheepeng** [comment]
@moziliar kind reminder on this PR.

**Comment by moziliar** [comment]
Sorry for leaving this PR hanging for so long; I might need some time to reconcile it with a recent change.

**Comment by madanalogy** [comment]
Closed in favour of #11567

**Pull request closed without merging**
